### PR TITLE
Update hncconfig webhook to handle object overwriting and refactor forest `sourceObjects` field

### PR DESCRIPTION
**Pull request by yiqigao217**
Update forest field `sourceObjects` to store all the names of user-created
objects by GVK as long as there's an object reconciler for that GVK.

Change object reonciler behaviors dependent on the previous forest field
`originalObjects`, which only stores source objects for 'Propagate' types.
Add test cases for corner cases.

Add new hncconfig webhook rules and new test cases. Check if changing
type mode(s) to `Propagate` would cause overwriting source by looking
up the `objectNames` data. Deny the request and output all the conflicts
if it would cause overwriting.

Tested by "make test" and manually on a GKE cluster with multiple
objects with the same name in the same tree or different trees. Passed
all e2e tests too.

The webhook message would look like this:
```
Cannot update configuration because setting type "/v1, Kind=Secret" to 'Propagate' mode would overwrite user-created object(s):
  Object "my-creds2" in namespace "acme-org" would overwrite the one in "team-a"
  Object "my-creds2" in namespace "acme-org" would overwrite the one in "team-b"
  Object "my-creds" in namespace "bcme-org" would overwrite the one in "team-c"
  Object "my-creds2" in namespace "bcme-org" would overwrite the one in "team-c"
To fix this, please rename or remove the conflicting objects first.
```
Part of #1102, #1076

### CODE REVIEW

File: incubator/hnc/internal/forest/namespaceobjects.go [hunk 1]
```diff
@@ -5,76 +5,77 @@ import (
...
-func (ns *Namespace) GetPropagatedObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
+// GetAncestorSourceObjects returns all source objects in the ancestors.
+func (ns *Namespace) GetAncestorSourceObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
 	o := []*unstructured.Unstructured{}
 	ans := ns.AncestryNames()
 	for _, n := range ans {
 		// Exclude the original objects in this namespace
 		if n == ns.name {
 			continue
 		}
```
Comment (adrianludwin) [inline incubator/hnc/internal/forest/namespaceobjects.go:61]: How about we remove lines 58-61, get rid of `GetAncestorAndLocalSourceObjects`, and then the (one current) caller to this function just says `ns.GetParent().GetAncestorSourceObjects()` instead of `ns.GetAncestorSourceObjects()`? You could change this function to return nil if `ns` is nil.

This would simplify the implementation and reduce the confusion between having _three_ similar functions instead of just two. Maybe at that point, you could have `GetLocalSourceObjects` and `GetAncestorSourceObjects`.

Also, [according to the Intro to Algorithms book](https://stackoverflow.com/a/3078231/158143), a node considered to be one of its own ancestors. So this is a defensible choice and also matches the behaviour of `ns.AncestryNames()` (maybe we should change that to `AncestorNames`).

File: incubator/hnc/internal/forest/namespaceobjects.go [hunk 2]
```diff
@@ -5,76 +5,77 @@ import (
...
 }
 
-// GetSource returns the original copy in the ancestors if it exists.
-// Otherwise, return nil.
-func (ns *Namespace) GetSource(gvk schema.GroupVersionKind, name string) *unstructured.Unstructured {
-	pos := ns.GetPropagatedObjects(gvk)
+// GetAncestorSourceObject returns the source object in the ancestors if it
+// exists. Otherwise, return nil.
+func (ns *Namespace) GetAncestorSourceObject(gvk schema.GroupVersionKind, name string) *unstructured.Unstructured {
+	pos := ns.GetAncestorSourceObjects(gvk)
```
Comment (adrianludwin) [inline incubator/hnc/internal/forest/namespaceobjects.go:78]: I don't think this is correct anymore because it doesn't account for the fact that the most recent source object with the same name might not have been propagated (and this will be even more important with exceptions). This should probably return a list, starting with any object in the most recent ancestor.

You could even combine this with `GetAncestorSourceObjects` (since this is no longer returning a single object anymore). If `name` is empty, it returns everything; otherwise it _only_ matches objects with the same name.

File: incubator/hnc/internal/reconcilers/object.go [hunk 3]
```diff
@@ -332,11 +333,13 @@ func (r *ObjectReconciler) syncObject(ctx context.Context, log logr.Logger, inst
...
 	// see if there's a conflicting source.
-	srcInst := r.Forest.Get(inst.GetNamespace()).GetSource(r.GVK, inst.GetName())
+	srcInst := r.Forest.Get(inst.GetNamespace()).GetAncestorSourceObject(r.GVK, inst.GetName())
 
 	// The object is a source without conflict if a copy of the source is not
-	// found in the forest or itself is found.
-	if srcInst == nil || srcInst.GetNamespace() == inst.GetNamespace() {
+	// found in the forest or itself is found. There won't be any conflict for
+	// types with sync modes other than 'Propagate' either.
+	mode := r.Forest.GetTypeSyncer(inst.GroupVersionKind()).GetMode()
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:341]: Can use `r.GVK` instead of `inst.GroupVersionKind()` here.

File: incubator/hnc/internal/reconcilers/object.go [hunk 4]
```diff
@@ -332,11 +333,13 @@ func (r *ObjectReconciler) syncObject(ctx context.Context, log logr.Logger, inst
...
+	if srcInst == nil || srcInst.GetNamespace() == inst.GetNamespace() || mode != api.Propagate {
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:342]: How could `srcInst.GetNamespace() == inst.GetNamespace()` today? Can't we get rid of this clause?

I also wonder if this code would now be written more naturally in the reverse (I had a bit of a hard time following the logic):

```
if srcInst != nil && mode == api.Propagate {
   // Since this is a conflict, overwrite
   return r.syncPropagated(ctx, log, inst)
}

// This object is not being overwritten; register the new source object and don't take any other action.
r.syncSource(ctx, log, inst)
return actionNop, nil
```

Or maybe even combine both calls in syncPropagated to one in this function by changing `hasPropagatedLabel` to `isPropagated`? That way you don't then have to "rediscover" the source object inside `syncPropagated` as we do today. E.g.:

```
if srcInst, isPropagated := r.isPropagated(inst); isPropagated {
  // srcInst can be nil if the source was removed
  return r.syncPropagated(ctx, log, srcInst, inst)
}
r.syncSource(ctx, log, inst)
return actionNop, nil
```
```

**Review by adrianludwin — COMMENTED** [review]
This is my review for the first commit only. Overall I really like it; I've left some comments that could simplify it even further (and possibly identifies one bug).

File: incubator/hnc/internal/validators/hncconfig.go [hunk 5]
```diff
@@ -141,6 +147,171 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+		return deny(metav1.StatusReasonConflict, printMsgWithConflicts(conflicts))
+	}
+	return allow("")
+}
+
+// checkConflictsForGVK checks object naming conflicts for a specific type.
+func (c *HNCConfig) checkConflictsForGVK(gvk schema.GroupVersionKind, conflicts conflicts) {
+	// We will keep one object set for each type, because objects with the same
+	// name but different types won't affect each other.
+	obs := objNameSet{}
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:186]: I find this logic quite hard to reason about, because we process the namespaces in a random order but maintain state between visiting each one.

As an alternative, how about this: do a _top-down_ traversal with the only information being passed from parent to child being the list of objects and the namespaces that they're from (since not all source objects are propagated, especially with exceptions, there can be multiple ancestors with objects with the same name). E.g. something like this:

```
// ancestorObjects maps an object name to the ancestor namespace(s) in which it's defined
type ancestorObjects map[string][]string
func (a ancestorObjects) copy() ancestorObject {... makes a copy ...}

func (...) checkConflictsForGVK(gvk) []string {
  conflicts := []string{}
  for _, ns := range c.Forest.GetRoots() {
    conflicts = append(conflicts, c.checkConflictsForNS(gvk, ancestorObjects{}, ns)
  }
}

func (...) checkConflictsForNS(gvk, ao, ns) []string {
  conflicts := []string{}
  objs := ao.copy()
  for _, o := ns.GetSourceObjects(gvk) {
    onm := o.GetName()
    if len(objs[onm]) > 0 {
      // TODO: check if this is a real conflict
      conflicts = append(conflicts, fmt.Sprintf("Object %q in namespace %q overwrites object in %q", onm, <todo>, ns.GetName())
      continue
    }
    objs[onm] = append(objs[onm], ns.GetName())
  }
  for _, cns := ns.GetChildren() {
    conflicts = append(conflicts, r.checkConflictsForNS(gvk, objs, cns)
  }
  return cns
}
```

File: incubator/hnc/internal/validators/hncconfig_test.go [hunk 6]
```diff
@@ -197,6 +201,56 @@ func TestNonRBACTypes(t *testing.T) {
 	}
 }
 
+func TestPropagateConflict(t *testing.T) {
+	// Create a forest with "a" as the parent and "b" and "c" as the children.
+	f := foresttest.Create("-aa")
+
+	tests := []struct {
+		name         string
+		inNamespaces []string
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig_test.go:210]: Since each namespace name is a single character, this can just be a string (that was on purpose 🙂 )

Although with that said - when we start testing exceptions, this code could become complicated enough that we'll need maps/lists anyway.

File: incubator/hnc/internal/validators/hncconfig_test.go [hunk 7]
```diff
@@ -197,6 +201,56 @@ func TestNonRBACTypes(t *testing.T) {
...
+			for _, nsn := range tc.inNamespaces {
+				f.Get(nsn).SetSourceObject(inst)
+			}
+			got := config.handle(context.Background(), c)
+
+			// Clean up source objects before next test case.
+			for _, nsn := range tc.inNamespaces {
+				f.Get(nsn).DeleteSourceObject(schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Secret"}, "my-creds")
+			}
+
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig_test.go:247]: If you create the forest inside the test, there's nothing to clean up.

File: incubator/hnc/internal/validators/hncconfig_test.go [hunk 8]
```diff
@@ -197,6 +201,56 @@ func TestNonRBACTypes(t *testing.T) {
...
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			g := NewGomegaWithT(t)
+			configs := []api.TypeSynchronizationSpec{
+				{APIVersion: "rbac.authorization.k8s.io/v1", Kind: "Role", Mode: "Propagate"},
+				{APIVersion: "rbac.authorization.k8s.io/v1", Kind: "RoleBinding", Mode: "Propagate"},
+				{APIVersion: "v1", Kind: "Secret", Mode: "Propagate"}}
+			c := &api.HNCConfiguration{Spec: api.HNCConfigurationSpec{Types: configs}}
+			c.Name = api.HNCConfigSingleton
+			v := fakeGVKValidator{"CronTab"}
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig_test.go:231]: CronTab isn't used in this test, you can just leave it out.

```
config := &HNCConfig{validator: fakeGVKValidator{}, Forest: f}
```

File: incubator/hnc/internal/forest/namespaceobjects.go (hunk 1 above)
Comment (yiqigao217) [reply incubator/hnc/internal/forest/namespaceobjects.go:61]: sgtm!

File: incubator/hnc/internal/validators/hncconfig.go [hunk 9]
```diff
@@ -141,6 +147,171 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
 
+// objNameSet is a set of namespaced objects with object names as the key. It
+// also stores if the object has been checked for naming conflict.
+type objNameSet map[string]inNamespace
+type inNamespace struct {
+	namespace string
+	checked   bool
+}
+
+// conflicts stores object naming conflicts by GVK, object name and groups of
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/internal/validators/hncconfig.go:158]: Agreed. Removed this and used the recursion below. It's much easier now.

File: incubator/hnc/internal/forest/namespaceobjects.go (hunk 2 above)
Comment (yiqigao217) [reply incubator/hnc/internal/forest/namespaceobjects.go]: True, I updated it with a list but still starting with the top ancestor since the reconciler behavior is to propagate the top "propagatable" ancestor (with no finalizer).

File: incubator/hnc/internal/reconcilers/object.go (hunk 3 above)
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object.go]: Updated all occurrences.

File: incubator/hnc/internal/reconcilers/object.go (hunk 4 above)
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object.go]: sgtm, reversed the logic!

File: incubator/hnc/internal/reconcilers/object_test.go [hunk 10]
```diff
@@ -191,6 +191,29 @@ var _ = Describe("Secret", func() {
 		Eventually(objectInheritedFrom(ctx, "Role", barName, "bar-role")).Should(Equal(fooName))
 	})
 
+	It("should overwrite conflicting source with the top source that can propagate", func() {
+		// Create a 'baz-role' in 'foo' that cannot propagate because of the finalizer.
+		makeObject(ctx, "Role", fooName, "baz-role")
+		Eventually(hasObject(ctx, "Role", fooName, "baz-role")).Should(BeTrue())
+		setFinalizer(ctx, fooName, "baz-role", true)
```
Comment (yiqigao217) [inline incubator/hnc/internal/reconcilers/object_test.go:198]: Added line 197 before `setFinalizer`. The previous flaky "baz-role" not found error should be fixed.

**Review by adrianludwin — COMMENTED** [review]
A few comments for the first commit - will come back and finish in a bit (just need to take a break)

File: incubator/hnc/internal/reconcilers/object.go [hunk 11]
```diff
@@ -325,34 +327,65 @@ func (r *ObjectReconciler) syncObject(ctx context.Context, log logr.Logger, inst
...
 	}
 
-	// Since there's a conflict that another source with the same name is found in
-	// the ancestors, this instance will be treated as propagated objects and will
-	// be overwritten by the source in the ancestor.
-	return r.syncPropagated(ctx, log, inst)
+	// If there's a conflicting source in the ancestors (excluding itself) and the
+	// the type has 'Propagate' mode, the object will be overwritten.
+	mode := r.Forest.GetTypeSyncer(r.GVK).GetMode()
+	if mode == api.Propagate && r.hasConflictingSource(log, inst) {
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:351]: Does it make sense to a) return the conflicting source here and b) return it from `shouldSyncAsPropagated` so we don't have to find this source a second time? I'm not worried about the runtime, it just seems a touch unfortunate to go through this logic twice.

But if it makes the code more convoluted, don't worry about it.

File: incubator/hnc/internal/reconcilers/object.go [hunk 12]
```diff
@@ -325,34 +327,65 @@ func (r *ObjectReconciler) syncObject(ctx context.Context, log logr.Logger, inst
...
+		return true
+	}
+
+	return false
+}
+
+// hasConflictingSource returns true if there's any conflicting source in the
+// ancestors that would cause overwriting this source object.
+func (r *ObjectReconciler) hasConflictingSource(log logr.Logger, inst *unstructured.Unstructured) bool {
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:360]: Should this just be:

```
return r.getTopSourceToPropagate(log, inst) != nil
```
?

File: incubator/hnc/internal/reconcilers/object.go [hunk 13]
```diff
@@ -382,9 +415,29 @@ func (r *ObjectReconciler) syncPropagated(ctx context.Context, log logr.Logger,
...
+		}
+		return obj
+	}
+	return nil
+}
+
 // syncSource syncs the copy in the forest with the current source object. If there's a change,
 // enqueue all the descendants to propagate the new source.
 func (r *ObjectReconciler) syncSource(ctx context.Context, log logr.Logger, src *unstructured.Unstructured) {
+	// Update or create a copy of the source object in the forest.
```
Comment (adrianludwin) [inline incubator/hnc/internal/reconcilers/object.go:437]: Maybe comment why we're doing this even if we're not propagating it.

File: incubator/hnc/internal/forest/namespaceobjects.go [hunk 14]
```diff
@@ -5,80 +5,76 @@ import (
...
-// GetPropagatedObjects returns all original copies in the ancestors.
-func (ns *Namespace) GetPropagatedObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
-	o := []*unstructured.Unstructured{}
+// GetAncestorSourceObjects returns all source objects with the specified name
+// in the ancestors (including itself) from top down. If the name is not
+// specified, all the source objects in the ancestors will be returned.
+func (ns *Namespace) GetAncestorSourceObjects(gvk schema.GroupVersionKind, name string) []*unstructured.Unstructured {
+	if ns == nil {
+		return nil
+	}
```
Comment (adrianludwin) [inline incubator/hnc/internal/forest/namespaceobjects.go:59]: Maybe comment why we have this here - eg because it allows you to say `ns.GetParent().GetAncestorSourceObjects()` without caring if the namespace is a root.

File: incubator/hnc/internal/validators/hierarchy.go [hunk 15]
```diff
@@ -246,15 +246,15 @@ func (v *Hierarchy) getConflictingObjectsOfType(gvk schema.GroupVersionKind, new
 	// Get all the source objects in the new ancestors that would be propagated
 	// into the descendants.
 	newAnsSrcObjs := make(map[string]bool)
-	for _, o := range newParent.GetPropagatingObjects(gvk) {
+	for _, o := range newParent.GetAncestorSourceObjects(gvk, "") {
 		newAnsSrcObjs[o.GetName()] = true
 	}
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hierarchy.go:251]: TODO (maybe for Ginny): this doesn't obey the shouldPropagate rules, but it probably should (and will definitely have to once we have exceptions).

**Review by adrianludwin — COMMENTED** [review]
Ok I'm now fully finished with the first commit :) On to the second commit

File: incubator/hnc/internal/forest/forest.go [hunk 16]
```diff
@@ -96,6 +96,17 @@ func (f *Forest) GetNamespaceNames() []string {
 	return names
 }
 
+// GetRoots returns all the root namespaces in the cluster.
```
Comment (adrianludwin) [inline incubator/hnc/internal/forest/forest.go:99]: Comment that any cycles will be omitted.

File: incubator/hnc/internal/validators/hncconfig.go [hunk 17]
```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+
+	// Get types that are changed from other modes to "Propagate" mode.
+	gvks := c.getNewPropagateTypes(inst)
+
+	// Check if user-created objects would be overwritten by these mode changes.
+	for gvk, _ := range gvks {
+		conflicts := c.checkConflictsForGVK(gvk)
+		if len(conflicts) != 0 {
+			msg := fmt.Sprintf("Cannot update configuration because setting type %q to 'Propagate' mode would overwrite user-created object(s):\n", gvk)
+			msg += strings.Join(c.checkConflictsForGVK(gvk), "\n")
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:162]: Should this just be `conflicts` instead of calling the function again?

File: incubator/hnc/internal/validators/hncconfig.go [hunk 18]
```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+
+// ancestorObjects maps an object name to the ancestor namespace(s) in which
+// it's defined.
+type ancestorObjects map[string][]string
+
+// checkConflictsForGVK looks for conflicts from top down for each tree.
+func (c *HNCConfig) checkConflictsForGVK(gvk schema.GroupVersionKind) []string {
+	conflicts := []string{}
+	for _, ns := range c.Forest.GetRoots() {
+		conflicts = append(conflicts, c.checkConflictsForNS(gvk, ancestorObjects{}, ns)...)
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:179]: Nit: consider checkConflictsForTree?

File: incubator/hnc/internal/validators/hncconfig.go [hunk 19]
```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+}
+
+func (a ancestorObjects) copy() ancestorObjects {
+	copy := ancestorObjects{}
+	for k, v := range a {
+		copy[k] = v
+	}
+	return copy
+}
+
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:234]: Nit: move this to just after where `ancestorObjects` is defined (or move `ancestorObjects` down here)

File: incubator/hnc/internal/validators/hncconfig.go [hunk 20]
```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+	}
+	return conflicts
+}
+
+func (c *HNCConfig) checkConflictsForNS(gvk schema.GroupVersionKind, ao ancestorObjects, ns *forest.Namespace) []string {
+	conflicts := []string{}
+	objs := ao.copy()
+	for _, o := range ns.GetSourceObjects(gvk) {
+		onm := o.GetName()
+		objs[onm] = append(objs[onm], ns.Name())
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:189]: Consider `objs.add(ns, onm)`?

File: incubator/hnc/internal/validators/hncconfig.go [hunk 21]
```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+		// If there are more than just the one (current) namespace we just added,
+		// there may be conflicts.
+		if len(objs[onm]) > 1 {
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:192]: Consider `objs.hasConflict(onm)`?

File: incubator/hnc/internal/validators/hncconfig.go [hunk 22]
```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+			// Currently the top conflicting ancestor would overwrite the object.
+			// TODO: check if this is a real conflict
+			conflicts = append(conflicts, fmt.Sprintf("  Object %q in namespace %q would overwrite the one in %q", onm, objs[onm][0], ns.Name()))
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:195]: Consider `objs.source(onm)` or `objs.top(onm)` or something like that instead of `objs[onm][0]`?

File: incubator/hnc/internal/validators/hncconfig.go [hunk 23]
```diff
@@ -141,6 +147,91 @@ func (c *HNCConfig) validateType(ctx context.Context, t api.TypeSynchronizationS
...
+		}
+	}
+	for _, cnm := range ns.ChildNames() {
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig.go:198]: Comment on why this can't cause a cycle :D

File: incubator/hnc/internal/validators/hncconfig_test.go [hunk 24]
```diff
@@ -48,7 +51,8 @@ func TestDeletingOtherObject(t *testing.T) {
 }
 
 func TestRBACTypes(t *testing.T) {
-	config := &HNCConfig{}
+	f := forest.NewForest()
+	config := &HNCConfig{Forest: f}
```
Comment (adrianludwin) [inline incubator/hnc/internal/validators/hncconfig_test.go:55]: Delete?

**Review by adrianludwin — COMMENTED** [review]
Second commit lgtm apart from some relatively minor suggestions

File: incubator/hnc/internal/reconcilers/object.go [hunk 25]
```diff
@@ -274,14 +274,16 @@ func (r *ObjectReconciler) skipNamespace(ctx context.Context, log logr.Logger, i
...
 		return actionUnknown
 	}
 
+	ns := r.Forest.Get(inst.GetNamespace())
 	// If it's a source, it must have been deleted. Update the forest and enqueue all its
 	// descendants, but there's nothing else to do.
-	if r.forestHasSource(inst) {
+	if ns.HasSourceObject(r.GVK, inst.GetName()) {
+		ns.DeleteSourceObject(r.GVK, inst.GetName())
 		r.syncUnpropagatedSource(ctx, log, inst)
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object.go:287]: It has reconcile HNCConfig and `enqueueDescendants()` to delete the obsolete copies.

File: incubator/hnc/internal/reconcilers/object.go (hunk 11 above)
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object.go:351]: Updated to return `srcInst`.

File: incubator/hnc/internal/reconcilers/object.go [hunk 26]
```diff
@@ -325,34 +327,65 @@ func (r *ObjectReconciler) syncObject(ctx context.Context, log logr.Logger, inst
...
 
-	// Since there's a conflict that another source with the same name is found in
-	// the ancestors, this instance will be treated as propagated objects and will
-	// be overwritten by the source in the ancestor.
-	return r.syncPropagated(ctx, log, inst)
+	// If there's a conflicting source in the ancestors (excluding itself) and the
+	// the type has 'Propagate' mode, the object will be overwritten.
+	mode := r.Forest.GetTypeSyncer(r.GVK).GetMode()
+	if mode == api.Propagate && r.hasConflictingSource(log, inst) {
+		return true
```
[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object.go:352]: sgtm updated.

File: incubator/hnc/internal/reconcilers/object.go (hunk 12 above)
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object.go:360]: Removed this dup function.

File: incubator/hnc/internal/reconcilers/object.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (adrianludwin) [reply incubator/hnc/internal/reconcilers/object.go:287]: No, I just mean inline its features into this function since it does so little.

[the remark this answers is not in the document]
Comment (yiqigao217) [reply incubator/hnc/internal/reconcilers/object.go:287]: `syncUnpropagatedSource` func is used twice and it contains two parts. I feel better to keep this logic stand alone, WDYT?
```
// syncUnpropagatedSource enqueues any propagated copies of the source.
//
// The method can be called when the source was deleted by users, or if it will no longer be
// propagated, e.g. because the user's changed the sync mode in the HNC Config.
func (r *ObjectReconciler) syncUnpropagatedSource(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) {
	// Signal the config reconciler for reconciliation because it is possible that the source object is
	// deleted on the apiserver.
	hnccrSingleton.requestReconcile("possible deleted source object")

	r.enqueueDescendants(ctx, log, inst, "possibly deleted source object")
}
```

File: incubator/hnc/internal/validators/hncconfig.go (hunk 17 above)
Comment (yiqigao217) [reply incubator/hnc/internal/validators/hncconfig.go:162]: Ah right.

File: incubator/hnc/internal/validators/hncconfig_test.go (hunk 24 above)
Comment (yiqigao217) [reply incubator/hnc/internal/validators/hncconfig_test.go:55]: It will have nil pointer error if we have forest as empty.

File: incubator/hnc/internal/reconcilers/object.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (adrianludwin) [reply incubator/hnc/internal/reconcilers/object.go:287]: Ah sgtm

**Review by adrianludwin — COMMENTED** [review]
/lgtm
/approve
/hold
/assign @rjbez17 

Hey Ryan, this is a big PR with two commits. I'd recommend reviewing them individually:
1. Commit #1 is about _always_ saving _all_ source objects in the forest, even if we're not propagating objects. This enables...
2. Commit #2 which checks the forest for objects that are about to be overwritten if we change the mode to Propagate.

Note that #2 will only have an effect if the reconciler is currently not in the Ignore mode for that type of object - which it typically will be. A future change will guide users to _first_ change to the Remove mode, and only _later_ change to the Propagate mode.

**Comment by yiqigao217** [comment]
Update: Passed all e2e tests.

Initially I got:
```
Summarizing 3 Failures:

1.[Fail] Demo [It] Should test basic functionalities in demo 
/usr/local/google/home/yiqigao/go/src/sigs.k8s.io/multi-tenancy/incubator/hnc/pkg/testutils/testutils.go:110

2. [Fail] HNC should delete and create a new Rolebinding instead of updating it [It] Should delete and create a Rolebinding when HNC is undeployed - issue #798 
/usr/local/google/home/yiqigao/go/src/sigs.k8s.io/multi-tenancy/incubator/hnc/pkg/testutils/testutils.go:110

3. [Fail] When deleting CRDs [It] should not delete subnamespaces 
/usr/local/google/home/yiqigao/go/src/sigs.k8s.io/multi-tenancy/incubator/hnc/pkg/testutils/testutils.go:69

Ran 29 of 29 Specs in 1030.354 seconds
FAIL! -- 26 Passed | 3 Failed | 0 Pending | 0 Skipped
```
1 & 2 passed when I ran it manually. 3 passed by itself if it's a stand-alone test (with `Fit`).

The 1st demo failure also failed on master. I suspect there's a bug in the test that it should use `RoleBinding` instead of `Role` here: https://github.com/kubernetes-sigs/multi-tenancy/blob/f990be3264cac6edb7618880b859522d25d40989/incubator/hnc/test/e2e/demo_test.go#L55

/assign @adrianludwin 
/cc @GinnyJI

**Comment by adrianludwin** [comment]
Yes, @GinnyJI is looking into the e2e test failures on `master`. I'm comfortable that this change is good and doesn't break anything.
